common/src/character_controller/vector_bounds.rs

```diff
@@ -0,0 +1,298 @@
+                warn!("Unsatisfied existing bound is parallel to new bound. Is the character squeezed between two walls?");
+                continue;
+            };
+
+            let mut candidate = *vector;
+            ortho_bound.constrain_vector(&mut candidate, self.error_margin);
+
+            if (self.bounds.iter()).all(|b| b.check_vector(&candidate, self.error_margin)) {
+                *vector = candidate;
+                if let Some(ref mut tagalong) = tagalong {
+                    ortho_bound.constrain_vector(tagalong, 0.0);
+                }
+                return;
+            }
+        }
+
+        // If no choice satisfies all constraints, keep all bounds and set the vector to 0
+        vector.set_zero();
+        if let Some(ref mut tagalong) = tagalong {
+            tagalong.set_zero();
+        }
+    }
+}
+
+/// Represents a single constraint for a vector. `VectorBound`s alone conceptually contain
+/// enough information to apply to a vector, but practically, one other piece of information
+/// is needed: `error_margin`, which exists in `VectorBoundGroup`.
+#[derive(Clone)]
+pub struct VectorBound {
+    normal: na::UnitVector3<f32>,
+    projection_direction: na::UnitVector3<f32>,
+    error_margin_factor: f32, // Margin of error when the bound is applied
+}
+
+impl VectorBound {
+    /// Creates a `VectorBound` that pushes vectors away from the plane given
+    /// by the normal in `projection_direction`. After applying such a bound to
+    /// a vector, its dot product with `normal` should be positive even counting
+    /// floating point approximation limitations.
+    pub fn new_push(
+        normal: na::UnitVector3<f32>,
+        projection_direction: na::UnitVector3<f32>,
+    ) -> Self {
+        VectorBound {
+            normal,
+            projection_direction,
+            error_margin_factor: 1.0,
+        }
+    }
+
+    /// Updates `subject` with a projection transformation based on the constraint given by `self`.
+    /// This function does not check whether such a constraint is needed.
+    fn constrain_vector(&self, subject: &mut na::Vector3<f32>, error_margin: f32) {
+        math::project_to_plane(
+            subject,
+            &self.normal,
+            &self.projection_direction,
+            error_margin * self.error_margin_factor,
+        );
+    }
+
+    /// Checks whether `subject` satisfies the constraint given by `self`. Note that `check_vector` will
+    /// return `true` after a vector is constrained by `constrain_vector` with the same error margin, even
+    /// if it's perturbed slightly. However, that property only holds if the error margin is not too small.
+    fn check_vector(&self, subject: &na::Vector3<f32>, error_margin: f32) -> bool {
+        // An additional margin of error is needed when the bound is checked to ensure that an
+        // applied bound always passes the check.
+        let error_margin_factor_for_check = self.error_margin_factor - 0.5;
```

Comment:
Why 0.5 specifically? That seems like a significant figure.

Reply:
This definitely deserves at least a comment. There's a pretty concrete explanation for when `error_margin_factor` is 1.0:
- If we used 0.0, the error margin would be the same when constraining and checking, which means that the check might fail when it should succeed.
- If we used 1.0, the error margin when checking would be 0, which means that the code might falsely believe everything is okay, even when the vector is not facing far enough away from a wall. This could cause the character to get stuck, as a collision could be detected if machine precision causes collision checking to think the character is going towards that wall.

In short, the error margin factor can't be too close to the constraining error margin factor, nor can it be too close to 0. Therefore, I picked the value halfway in-between.

It might make more sense to use 2 and 1 instead of 1 and 0.5 (for error margin factor and the check delta).